### ISSUE TITLE
Http response custom header not working properly asp.net core 2.2 web api response.

### ISSUE THREAD

**Issue by Eilon** [dotnet/aspnetcore#8567]
_From @Gopigovin on Thursday, 14 March 2019 10:54:34_

Hi Guys, 

Http response custom header not working properly asp.net core 2.2 web api response.

We used the html upload box in our project. While upload fails, we will pass the custom message through httprequest to client side. Our server side section is in .net core project (2.1). So we have used the below code to send custom message

Response.Headers.Add("status", "File processed correctly");

It working fine core 2.1. While using the above code in .net core 2.2 project, it does not works.

How to resolve this in our side? Please suggest your ideas.

Regards,
Gopi G.

_Copied from original issue: dotnet/core#2456_

**Comment by Eilon** [comment]
_From @leecow on Friday, 15 March 2019 17:38:27_

@dougbu @Eilon

**Comment by Eilon** [comment]
_From @dougbu on Friday, 15 March 2019 17:52:12_

@Tratcher @anurse @muratg any ideas?

**Comment by Eilon** [comment]
@Gopigovin it would be helpful if you could share a complete application that reproduces this problem.

**Comment by jkotalik** [comment]
Also, do you happen to know which server type you are using?

**Comment by Gopigovin** [comment]
Hi @Eilon , 

We have prepared sample with html upload with asp.net core(2.2) web api service. 

1. Extract the sample - [backend.zip](https://github.com/aspnet/AspNetCore/files/2977175/backend.zip)

2. open the htmlpage.html file into browser. 
3. Run the backend.sl file in Visual Studio. 
4. Upload any file and check the onload() event, cannot able to get the custom header text(**"status"**)

Please provide solution for this query. 

Regards,
Gopi G.

**Comment by Gopigovin** [comment]
Hi @jkotalik , 

We have used Web API service into ASP.Net Core 2.2. Please find the sample and replication procedure. 

Extract the sample - backend.zip

open the htmlpage.html file into browser.

Run the backend.sl file in Visual Studio.

Upload any file and check the onload() event, cannot able to get the custom header text("status")

Please provide solution for this query.

Regards,
Gopi G.

**Comment by Tratcher** [comment]
With your repro I see the value `status: 200, File processed correctly` in Chrome's F12 debugger tools. I think that part of the issue here is that you're getting HTTP/2 over HTTPS by default, and in HTTP/2 `:status:` is a reserved header where the status code `200` goes. Not sure why chrome is combining that with the custom header for the display.

Does it work as expected if you use a header name other than `status`?

**Comment by Gopigovin** [comment]
Hi @Tratcher , 

We have already tried to different name and change the header name ("**status - custom**"). But not working, the custom headers show in chrome browser network tab, not return the all headers in **getAllResponseHeaders()** method. please find the screen shot for your reference. 

![image](https://user-images.githubusercontent.com/17155380/54735837-46855480-4bce-11e9-87db-84ebec8a4e33.png)

Modified sample : 
[backend.zip](https://github.com/aspnet/AspNetCore/files/2991004/backend.zip)

Any other way to achieve this?

Regards,
Gopi G.

**Comment by Tratcher** [comment]
Then that's a usage issue with getAllResponseHeaders. It looks like it's related to CORS.
https://stackoverflow.com/a/7463297/2588374

**Comment by thiyagu-subramani** [comment]
Hi @Tratcher ,

We are tried the provided solution from the stack over flow link. But this is not working for us while disable the cross origin request in chrome. This issue is raised not only in the Core 2.2 application. This issue occur in dot net core 2.0 application also. So, can you please provide any other possible solution?



Regards,
Thiyagu S

**Comment by Tratcher** [comment]
@rynowak can you redirect for advise on CORS?

**Comment by rynowak** [comment]
CORS requires you to specify on the policy which headers you want to allow clients to see. https://github.com/aspnet/AspNetCore/blob/master/src/Middleware/CORS/src/Infrastructure/CorsPolicyBuilder.cs#L108

**Comment by JakeWierzba** [comment]
Has anyone had any luck getting this working?
I don't believe it its CORS (on the dotnet side anyway) as I have AllowAnyHeader() set on my policy.

**Comment by NetTecture** [comment]
It isn ot cors - someone did not read the documentation. The Cors AllowAnyHeaders is about headers the client can send, not headers returned to the client.

**Comment by pandiyarajm93** [comment]
I am also facing same issue with ASP.NET Core 3.1 Web API project (Server) with Blazor project (Client). Have any solution to this?

**Comment by MaximeMorin** [comment]
I was also facing a similar issue where my custom header wouldn't be sent from the controller when added like so:

```
this.Response.Headers.Add("my-header", "my-header-value");
```

But when adding as follow:
```
var headers = this.ControllerContext.HttpContext.Response.GetTypedHeaders();
headers.Append("my-header", "my-header-value");
```
It works. 

To be honest, I'm not sure what's going on.  I believe this is cause by the fact that I return a file.

```
return this.File(byteArray, "application/octet-stream", name);
```

For the solution, I got inspired FileResultExecutorBase and how they use the headers. 
https://github.com/dotnet/aspnetcore/blob/c565386a3ed135560bc2e9017aa54a950b4e35dd/src/Mvc/Mvc.Core/src/Infrastructure/FileResultExecutorBase.cs

Note: I'm also using CORS for development purposes so I had to add my header to the list of exposed headers.

```
app.UseCors(options => options.AllowAnyOrigin().AllowAnyMethod().AllowAnyHeader().WithExposedHeaders("my-header"));
```
